### PR TITLE
[feat] CI testing kubernetes version matrix

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

Enables tests for `main` and `next` pushes to run tests for the KIC against a history of Kubernetes versions.

**Which issue this PR fixes**

Fixes https://github.com/Kong/kubernetes-ingress-controller/issues/1542

### CODE REVIEW

**Comment by shaneutt** [comment]
kic-version-tests

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (mflendrich) [inline .github/workflows/2x-kubernetes-version-test.yaml:10]: See how go-kong achieves multiversion testing without code duplication:

https://github.com/Kong/go-kong/blob/100d6b60b97942ccaa2c5c319fd72bad53bc9341/.github/workflows/integration-test.yaml#L9

I think that something similar should be doable here - for dbless/postgres as well as for k8s versions.

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [inline railgun/test/integration/suite_test.go:90]: style inconsistent with the variable `dbmode` above.

The indirection with `ClusterVersionEnvVar` seems superfluous to me but that's only my preference. Also, the env var name appears unnecesarily exported.

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [inline railgun/test/integration/suite_test.go:123]: another style inconsistency (inline `os.Getenv` vs global `os.Getenv`)

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [inline railgun/test/integration/suite_test.go:125]:
```suggestion
			fmt.Fprintf(os.Stderr, "Error: cant use both %s and KIND_CLUSTER at the same time", ClusterVersionEnvVar)
```

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [inline railgun/test/integration/suite_test.go:147]: when using semantic exit statuses, consider defining them in one place. Otherwise a future editor will need to grep for `33`, `34`, `35` etc to find the next free status code (and can fail)

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [inline railgun/test/integration/suite_test.go:205]:
```suggestion
		fmt.Fprintf(os.Stderr, "Error: could not retrieve server version for cluster: %w", err)
```

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [inline railgun/test/integration/suite_test.go:141]: Why do some prints go to stderr, and some to stdout?

Maybe we should we use stderr only? Note that stdout can be buffered which is a downside.

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (ccfishk) [inline .github/workflows/2x-kubernetes-version-test.yaml:36]: We can retrieve version using runtime.Version(), or corresponding env environment variables.

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (shaneutt) [reply .github/workflows/2x-kubernetes-version-test.yaml:10]: Sounds good, I'll switch to using that strategy instead.

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (shaneutt) [reply .github/workflows/2x-kubernetes-version-test.yaml:36]: I'm not sure what the relevance is here: this option tells the `setup-go` action which versions of Go we accept for builds (e.g. latest `1.16.x` ultimately).

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go:90]: Sounds good, let's tear it out:

```suggestion
	// clusterVersion indicates the version of Kubernetes to use for the tests (if the cluster was not provided by the caller)
	clusterVersionStr = os.Getenv("KONG_CLUSTER_VERSION")

```

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go:123]: historical, but since I'm changing other things anyway I'll change this.

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go:141]: All prints before exiting the program due to an error should be going to os.Stderr, if that's not the case then it's a typo. I'll review this and fix any problems.

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go:147]: Yeah, I've been thinking about this over and over and I just hadn't gotten around to it yet. Now is a good a time as any.

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go:205]: The `%w` doesn't make sense in this context: there's no caller to handle the error nevertheless unwrap it.

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (shaneutt) [reply .github/workflows/2x-kubernetes-version-test.yaml]: Well... this does end up making the status checks a lot more verbose but since this wont run on every PR anyway I think that'll be just fine.

Resolved with 4b02baa6ede4b68a32706f260dd31dec9728f142 :+1:

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go]: 0b12a0a596b609caf7e97222bcc46c5dc04a5c17 resolves this, but also takes it a step further and starts storing env overrides in one consistent grouping.

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go]: Resolved via c5969ec7d94cb399a9067c651faebc4395524f19

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go]: There are three uses of `os.Stdout` in the file:

- `278,291`: simply ensuring that any stdout that was produced before an error occurred doesn't get swallowed by the error, but sill emitting to the same  stream for consistency
- `336`: this one was a typo from weeks ago, but since I'm here I fixed it: 0f2ad268c563f3ed3e261e531dc604b75f556607

**Comment by shaneutt** [comment]
This is ready for re-review.

**Comment by shaneutt** [comment]
Tests were successful with the most recent update, rebasing and removing the pull request shim.

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (ccfishk) [reply .github/workflows/2x-kubernetes-version-test.yaml]: shouldn't you just read go-lang version from the environment instead of hardcode ?

File: railgun/test/integration/suite_test.go
Comment (ccfishk) [reply railgun/test/integration/suite_test.go]: not a blocker

File: .github/workflows/2x-kubernetes-version-matrix-test.yaml
Comment (mflendrich) [inline .github/workflows/2x-kubernetes-version-matrix-test.yaml:13]: Can we make `dbless`/`postgres` matrix fields as well? I don't see a reason for duplication here, but maybe I haven't read this diligently enough.

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [inline railgun/test/integration/suite_test.go:216]: nit: if format printing a type implementing `fmt.Stringer`, use `%v` on the type directly instead of `%s` on `x.String()`

**Review by mflendrich — APPROVED** [review]
conditional lgtm if both suggestions get applied

File: railgun/test/integration/suite_test.go
Comment (mflendrich) [reply railgun/test/integration/suite_test.go]: It's a style thing: `%w` is equivalent to `%v` per `fmt` spec and my style preference is that using `%v` is better than `%s` + `String()`/`Error()` for brevity.

Optional.

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (mflendrich) [reply .github/workflows/2x-kubernetes-version-test.yaml]: @ccfishk, this line sets up Go for the build process. Before this, there is no Go compiler installed whatsoever. Here we specify what version of Go will be available on the worker.

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [inline railgun/test/integration/suite_test.go:138]:
```suggestion
			fmt.Fprintf(os.Stderr, "Error: could not use existing cluster for test env: %w", err)
```

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go]: It's funny how often these formatting directives come up in PR reviews :joy: 

Ultimately, changing this will have basically no effect on the result, however if we can start just doing things one standard way that could be helpful to reduce these kinds of things in PR review so I went ahead and changed it to try and move towards a conformance of style as you suggest:

https://github.com/Kong/kubernetes-ingress-controller/pull/1558/commits/f7aa06e20578872cefd4eb743b93228e3faec0c9

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go:216]:
```suggestion
	fmt.Printf("INFO: testing environment is ready KUBERNETES_VERSION=(%v): running tests\n", serverVersion)
```

File: .github/workflows/2x-kubernetes-version-test.yaml
Comment (shaneutt) [reply .github/workflows/2x-kubernetes-version-test.yaml]: Please `unresolve` if there's any follow up here, but as Michal stated: there's no Golang available _before_ this step and this line simply informs which Golang to install.

Comment (ccfishk) [reply .github/workflows/2x-kubernetes-version-test.yaml]: @shaneutt  understand this is the first line we set the expected golang version. It is good. 👍🏻

Comment (ccfishk) [reply .github/workflows/2x-kubernetes-version-test.yaml]: @mflendrich thanks for pointing this out.

File: railgun/test/integration/suite_test.go
Comment (shaneutt) [reply railgun/test/integration/suite_test.go]: So as it turns out, trying to use `%w` in `Fprintf` will not compile:

```
test/integration/suite_test.go:138:4: Fprintf call has error-wrapping directive %w
```

We literally _can't_ do this.

**Comment by shaneutt** [comment]
double checking the new matrix configuration, will disable (drop the commit) before merging

File: .github/workflows/2x-kubernetes-version-matrix-test.yaml
Comment (shaneutt) [reply .github/workflows/2x-kubernetes-version-matrix-test.yaml:16]: Checked over [the documentation](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) and got this patched. I don't think it looks as nice:

Screenshot_2021-07-22_08-37-50

But it's functional and `does-the-right-thing.go` so :+1:

**Pull request merged**
